### PR TITLE
CSHARP-2963: And filter with $near translates to invalid MongoDb query.

### PR DESCRIPTION
**Pull request by DmitryLukyanov**
EG: https://evergreen.mongodb.com/version/5e94ea34e3c3312519b63a7a

### CODE REVIEW

File: src/MongoDB.Driver/FilterDefinitionBuilder.cs [hunk 1]
```diff
@@ -1493,6 +1493,16 @@ public FilterDefinition<TDocument> Where(Expression<Func<TDocument, bool>> expre
 
     internal sealed class AndFilterDefinition<TDocument> : FilterDefinition<TDocument>
     {
+        #region static
+        private static readonly string[] __filtersThatAllowOnlyDollarForm = new[]
+        {
+            "$geoWithin",
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver/FilterDefinitionBuilder.cs:1499]: the first two are mentioned in the ticket. I also checked the 3rd and 4th and they have the same behavior. As I understand there are no more geo query selectors (based on this https://docs.mongodb.com/manual/reference/operator/query-geospatial/)

Comment (rstam) [reply src/MongoDB.Driver/FilterDefinitionBuilder.cs:1499]: Please rename this variable to:

    __operatorsThatCantBeCombined

File: src/MongoDB.Driver/FilterDefinitionBuilder.cs [hunk 2]
```diff
@@ -1541,11 +1551,11 @@ private static void AddClause(BsonDocument document, BsonElement clause)
             else if (document.Contains(clause.Name))
             {
                 var existingClause = document.GetElement(clause.Name);
-                if (existingClause.Value is BsonDocument && clause.Value is BsonDocument)
+                if (existingClause.Value is BsonDocument existingClauseValue && clause.Value is BsonDocument clauseValue)
                 {
-                    var clauseValue = (BsonDocument)clause.Value;
-                    var existingClauseValue = (BsonDocument)existingClause.Value;
-                    if (clauseValue.Names.Any(op => existingClauseValue.Contains(op)))
+                    var topLevelElementName = clauseValue.ElementCount > 0 ? clauseValue.GetElement(0).Name : null;
```
Comment (rstam) [inline src/MongoDB.Driver/FilterDefinitionBuilder.cs]: Please rename this variable:

    clauseOperator

Comment (rstam) [reply src/MongoDB.Driver/FilterDefinitionBuilder.cs]: Question: how can `clauseValue.ElementCount` ever be 0?

File: src/MongoDB.Driver/FilterDefinitionBuilder.cs [hunk 3]
```diff
@@ -1541,11 +1551,11 @@ private static void AddClause(BsonDocument document, BsonElement clause)
...
+                    if (__filtersThatAllowOnlyDollarForm.Contains(topLevelElementName) ||
+                        clauseValue.Names.Any(op => existingClauseValue.Contains(op)))
```
Comment (rstam) [inline src/MongoDB.Driver/FilterDefinitionBuilder.cs]: Would you be willing to reverse the clauses so that we have the original first and the new one second?

     if (clauseValue.Names.Any(op => existingClauseValue.Contains(op)) ||
         __operatorsThatCantBeCombined.Contains(clauseOperator))

**Review by rstam — CHANGES_REQUESTED** [review]
Looks good. See minor change requests.

File: src/MongoDB.Driver/FilterDefinitionBuilder.cs (hunk 1 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/FilterDefinitionBuilder.cs:1499]: done

File: src/MongoDB.Driver/FilterDefinitionBuilder.cs (hunk 3 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/FilterDefinitionBuilder.cs]: done

File: src/MongoDB.Driver/FilterDefinitionBuilder.cs (hunk 2 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/FilterDefinitionBuilder.cs]: it's possible in the case when operator body is empty. See this test: https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/130/files#diff-9c6d64d11defdd6e9847744a1776c574R96

File: tests/MongoDB.Driver.Tests/FilterDefinitionBuilderTests.cs [hunk 4]
```diff
@@ -78,6 +78,32 @@ public void And_with_clashing_keys_but_different_operators_should_get_merged()
...
+        [Fact]
+        public void And_with_clashing_keys_and_different_operators_but_with_filters_that_support_only_dollar_form_and_empty_filter_should_ignore_empty_filter()
+        {
+            var subject = CreateSubject<BsonDocument>();
+
+            var combinedFilter = subject.And(
+                "{ geoField : { $near : [ 5.0, 6.0 ] } }",
+                "{ geoField : { } }");
+
+            Assert(combinedFilter, "{ geoField : { $near : [ 5.0, 6.0 ] } }");
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Tests/FilterDefinitionBuilderTests.cs:104]: NOTE: In case if the order of `And` operators will be inversed (the operator with an empty body will be the first operator), the resulted query will be in dollar form. This behavior is not related to this ticket, so I didn't touch it

**Comment by DmitryLukyanov** [comment]
ready for review

**Comment by DmitryLukyanov** [comment]
EG: https://evergreen.mongodb.com/version/5e962d97a4cf475aa0ea08fa

**Review by rstam — APPROVED** [review]
LGTM

File: src/MongoDB.Driver/FilterDefinitionBuilder.cs (hunk 1 above)
Comment (vincentkam) [reply src/MongoDB.Driver/FilterDefinitionBuilder.cs:1499]: Nit: Should this be `__operatorsThatCannotBeCombined`? We typically seem to spell out contractions.

File: tests/MongoDB.Driver.Tests/FilterDefinitionBuilderTests.cs [hunk 5]
```diff
@@ -78,6 +78,32 @@ public void And_with_clashing_keys_but_different_operators_should_get_merged()
...
+        [InlineData("{ geoField : { $near : [ 5.0, 6.0 ] } }", "{ geoField : { $geoWithin : { $box : [ [ 1.0, 2.0 ], [ 3.0, 4.0 ] ] } } }")]
+        [InlineData("{ geoField : { $nearSphere : { $geometry : { type : 'Point', coordinates : [ 1, 2 ] } } } }", "{ geoField : { $geoIntersects : { $geometry : { type : 'Polygon', coordinates: [ [ [ 1, 2 ], [ 3, 4 ], [ 5, 6 ], [ 7, 8 ] ] ] } } } }")]
+        [InlineData("{ geoField : { $geoIntersects : { $geometry : { type : 'Polygon', coordinates: [ [ [ 1, 2 ], [ 3, 4 ], [ 5, 6 ], [ 7, 8 ] ] ] } } } }", "{ geoField : { $nearSphere : { $geometry : { type : 'Point', coordinates : [ 1, 2 ] } } } }")]
+        public void And_with_clashing_keys_and_different_operators_but_with_filters_that_support_only_dollar_form_should_get_promoted_to_dollar_form(string firstFilter, string secondFilter)
+        {
+            var subject = CreateSubject<BsonDocument>();
+
+            var combinedFilter = subject.And(firstFilter, secondFilter);
+
+            Assert(combinedFilter, $"{{ $and : [ {firstFilter}, {secondFilter} ] }}");
```
Comment (vincentkam) [inline tests/MongoDB.Driver.Tests/FilterDefinitionBuilderTests.cs:92]: Nit: Should we use fluent assertions here?

**Review by vincentkam — APPROVED** [review]
Two nits, otherwise LGTM! 😸ᓚᘏᗢ

File: tests/MongoDB.Driver.Tests/FilterDefinitionBuilderTests.cs (hunk 4 above)
Comment (vincentkam) [reply tests/MongoDB.Driver.Tests/FilterDefinitionBuilderTests.cs:104]: Ditto the above nit about possibly using fluent assertions?  
Edit: Oops, n/m :)

File: tests/MongoDB.Driver.Tests/FilterDefinitionBuilderTests.cs (hunk 5 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/FilterDefinitionBuilderTests.cs:92]: in general yes, but this is the custom method which is defined in this class. So, in this particular case - no :)

File: src/MongoDB.Driver/FilterDefinitionBuilder.cs (hunk 1 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/FilterDefinitionBuilder.cs:1499]: done

File: tests/MongoDB.Driver.Tests/FilterDefinitionBuilderTests.cs (hunk 5 above)
Comment (vincentkam) [reply tests/MongoDB.Driver.Tests/FilterDefinitionBuilderTests.cs:92]: Whoops! Thanks for the explanation! :)
